parse/src/main/java/com/parse/ParseRESTCommand.java

```diff
@@ -215,6 +216,20 @@ protected void addAdditionalHeaders(ParseHttpRequest.Builder requestBuilder) {
         if (masterKey != null) {
             requestBuilder.addHeader(HEADER_MASTER_KEY, masterKey);
         }
+        try {
+            JSONObject jsonObject = jsonParameters != null ? new JSONObject(jsonParameters.toString()) : new JSONObject();
+            // using header names so we don't override a parameter with the same key name,
+            // using all headers to insure the requestId generated doesn't conflict with the rest of the users
+            if (installationId != null)
+                jsonObject.put(HEADER_INSTALLATION_ID, installationId);
+            if (sessionToken != null)
+                jsonObject.put(HEADER_SESSION_TOKEN, sessionToken);
+            if (masterKey != null)
+                jsonObject.put(HEADER_MASTER_KEY, masterKey);
+            requestBuilder.addHeader(HEADER_REQUEST_ID, ParseDigestUtils.md5(toDeterministicString(jsonObject)));
+        } catch (JSONException e) {
```

Comment:
The OS should provide a UUID API that takes care of seeding. Could you use the same logic for generating the UUID as we use in the other SDKs? For example the iOS SDK and JS SDK?

Reply:
I have rewritten the logic to use a UUID, as this is uniformed between SDKs.

But imo why isn't the UUID data based?, like this if the SDK calls ParseObject.save() it changes the ParseRequest which creates a new UUID, while using a data based UUID calling save won't change the UUID header.

Currently if I want the SDK to create an idempotent request I'd still set the HttpClient to use a longer timeout and increase the default max retries of requests.

Reply:
I am not very good at JS, but could you direct me to where the iOS SDK impliments the logic.

Reply:
> But imo why isn't the UUID data based?

It doesn't need to be based on object data. A UUID is a Universally **Unique** IDentifier, which ensures that every generated UUID is unique.

The part you'd have to take care about in the PR is that for a `ParseObject.save` that UUID is only generated one time and then reused if the request fails and the SDK auto-retries the request. Obviously, if you generated a new UUID with ever retry of the same request, then this wouldn't work.

Reply:
The current approach works as follows if all requests were unsuccessfull:

`ParseObject object = new ParseObject("GameScore");`
// calls a request and retries according to the max retries (default is 4) set during initilization, with the same requestId header. (5 calls with the same header)
`object.save();`

// this second method call... calls a request and retries according to the max retries set, with the same requestId header but different than the first save() method call.
`object.save();`

Reply:
The second `save` call should use a different UUID for the request ID. In fact, every `save` call should use its own UUID. Only auto-retries of each `save` call should use the same UUID.

Reply:
Yes, It is using exactly that. My explanation was kind of confusing.

I'll add a second test using the code in my previous comment.